tests/unit/Request.ts

```diff
@@ -0,0 +1,391 @@
+	'test-local-header-a': 'true',
+	'test-local-header-b': 'yes',
+	'test-override': 'overridden'
+};
+
+function runCollectionTest<T>(collection: dstore.Collection<T>, rangeArgs: dstore.FetchRangeArgs | RequestOptions, expected?: RequestOptions) {
+	const expectedResults: { [ name: string ]: number | string }[] = [
+		{ id: 1, name: 'one' },
+		{ id: 2, name: 'two' }
+	];
+	mockRequestProvider.setResponseText(JSON.stringify(expectedResults));
+	let resultsPromise: dstore.FetchPromise<any>;
+	if (!expected) {
+		expected = <RequestOptions> rangeArgs;
+		resultsPromise = collection.fetch();
+	}
+	else {
+		const { start, end } = <dstore.FetchRangeArgs> rangeArgs;
+		mockRequestProvider.setResponseHeaders({
+			'Content-Range': start + '-' + end + '/' + expectedResults.length
+		});
+		resultsPromise = collection.fetchRange(<dstore.FetchRangeArgs> rangeArgs);
+	}
+	return resultsPromise.then(function (results) {
+		expected.headers && mockRequestProvider.assertRequestHeaders(expected.headers);
+		expected.query && mockRequestProvider.assertQuery(new UrlSearchParams(expected.query));
+
+		// We cannot just assert deepEqual with results and expectedResults
+		// because the store converts results into model instances with additional members.
+		assert.strictEqual(results.length, expectedResults.length);
+		for (let i = 0; i < results.length; ++i) {
+			const result: { [ name: string ]: number | string } = results[i];
+			const expectedResult: { [ name: string ]: number | string } = expectedResults[i];
+			for (let key in expectedResult) {
+				assert.strictEqual(result[key], expectedResult[key]);
+			}
+		}
+		if (resultsPromise.totalLength instanceof Promise) {
+			return (<Promise<number>> resultsPromise.totalLength).then(function (totalLength) {
+				assert.strictEqual(totalLength, expectedResults.length);
+			});
+		}
+	});
+}
+
+let registryHandle: Handle;
+let treeTestRootData: string;
+function createRequestTests(Store: new (options?: RequestStoreArgs) => ConcreteRequest<any>) {
+	return {
+		name: 'dstore Request',
+
+		before: function () {
+			registryHandle = providerRegistry.register(/.*mockRequest.*/, mockRequestProvider.respond);
+			return request.get((<any> require).toUrl('tests/unit/data/treeTestRoot')).then(function (response: any) {
+				treeTestRootData = response.data;
+			});
+		},
+
+		after: function () {
+			registryHandle.destroy();
+		},
+
+		beforeEach: function () {
+			mockRequestProvider.setResponseText('{}');
+			mockRequestProvider.setResponseHeaders({});
+			store = new Store({
+				target: '/mockRequest/',
+				headers: globalHeaders,
+				Model: Model
+			});
+			store.Model.prototype.describe = function() {
+				return "name is " + this.name;
+			}
+		},
+
+		'filter': function () {
+			mockRequestProvider.setResponseText(treeTestRootData);
+
+			return store.filter<{ name: string; describe: () => string; someProperty: string }>('data/treeTestRoot').fetch().then(function (results) {
```

Comment:
Might make sense to define an internal interface in this file with `describe` and `someProperty?` for this and the next test.